### PR TITLE
Add override keyword for ensuring function overrides

### PR DESCRIPTION
**Pull request by federicobond**
Part of #2563.

### CODE REVIEW

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 1]
```diff
@@ -226,6 +226,10 @@ bool SyntaxChecker::visit(FunctionDefinition const& _function)
 				"\"."
 			);
 	}
+
+	if (v050)
+		const_cast<FunctionDefinition &>(_function).optionallyMarkAsOverride();
```
Comment (federicobond) [inline libsolidity/analysis/SyntaxChecker.cpp]: This is super ugly, but I couldn't come up with a better approach. We could clean this up for 0.5.0.

File: Changelog.md [hunk 2]
```diff
@@ -24,6 +25,7 @@ Features:
  * Interface: Provide ability to select target EVM version (homestead or byzantium, with byzantium being the default).
  * Standard JSON: Reject badly formatted invalid JSON inputs.
  * Type Checker: Disallow uninitialized storage pointers as experimental 0.5.0 feature.
+ * Support and recommend using ``emit EventName();`` to call events explicitly.
```
Comment (chriseth) [inline Changelog.md]: This looks like a rebase problem.

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 3]
```diff
@@ -57,6 +57,22 @@ bool StaticAnalyzer::visit(FunctionDefinition const& _function)
 	solAssert(m_localVarUseCount.empty(), "");
 	m_nonPayablePublic = _function.isPublic() && !_function.isPayable();
 	m_constructor = _function.isConstructor();
+
+	if (m_currentContract->sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::V050))
+	{
+		if (_function.isOverride() && !_function.annotation().superFunction)
+			m_errorReporter.declarationError(
+				_function.location(),
+				"Function signature does not override any function defined in parent contracts."
```
Comment (chriseth) [inline libsolidity/analysis/StaticAnalyzer.cpp]: I think we call them "base contracts".

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 4]
```diff
@@ -57,6 +57,22 @@ bool StaticAnalyzer::visit(FunctionDefinition const& _function)
...
+			);
+
+		else if (!_function.isOverride() && _function.annotation().superFunction)
+			m_errorReporter.declarationError(
+				_function.location(),
+				"Function overrides a parent function but is missing an override specifier."
```
Comment (chriseth) [inline libsolidity/analysis/StaticAnalyzer.cpp]: Perhaps just "overrides a function"?

Comment (chriseth) [reply libsolidity/analysis/StaticAnalyzer.cpp]: Can we find out the name of the contract the overridden function is defined in? Perhaps we should even specify a secondary source location.

File: libsolidity/ast/AST.cpp [hunk 5]
```diff
@@ -348,6 +348,23 @@ string FunctionDefinition::fullyQualifiedName() const
 	return sourceUnitName() + ":" + contract->name() + "." + fname;
 }
 
+void FunctionDefinition::optionallyMarkAsOverride()
+{
+	auto pos = find_if(
+		m_functionModifiers.begin(),
+		m_functionModifiers.end(),
+		[](ASTPointer<ModifierInvocation> modifier) {
```
Comment (chriseth) [inline libsolidity/ast/AST.cpp]: Please use `const&` here

**Review by chriseth — CHANGES_REQUESTED** [review]
The use of `superFunction` is not 100% correct. superFunction points to the next higher up function of the same name in the list of linearized base contracts. See the following example:
```
contract A {
    function f() {}
}
contract B {
    function f() {}
}
contract C is A, B {
}
```

When we compile `C`, then `B.f`'s `superFunction` is set, while it is not set when we compile `B`.

It somehow already sounds like this is the correct behaviour, because we do not want C to be compilable, but the error would be different.

Could you please experiment a little with multiple inheritance and write tests for that?

The combinations I can currently think of are:

 - the example above
 - the example above plus implementation of `f` in C with `override`
 - the same without `override`.
 - the example above plus that A and B inherit from the same base class (empty)
 - the same but with `f` added to that common base class
 - the same but with `f` also added to `C`

**Comment by axic** [comment]
Needs to be rebased.

**Comment by federicobond** [comment]
Rebased but I am not likely to continue to work on this soon. I welcome anyone to take it

**Comment by chriseth** [comment]
Do we want to take that in scope for 0.5.0 or not?

**Comment by ekpyron** [comment]
I think enforcing the use of ``override`` is a good idea and vote for it.

**Comment by federicobond** [comment]
How would we solve the problem described by @chriseth above? Should we disallow compilation of C?

**Comment by chriseth** [comment]
First suggestion: In that situation it only compiles if the most derived contract contains (or inherits) a function called `f` with a modifier `override(A,B)`. So the general rule would be:

A function that overrides a function of the same name and parameter types from a base contract `BaseContract` needs to either have a `override` or `override(BaseContract)` modifier. If a contract `C` inherits from multiple contracts which contain functions of the same name and parameter types, then `C` also has to define such a function with the modifier `override(A,B, ..., )`.

There is a special case where `C` inherits from both `A` and `B`; and `A` also inherits from `B` - not sure if `override(A)` would be sufficient is that case, but I think it is better to be explicit and require both `A` and `B`.

**Comment by chriseth** [comment]
Closing as there are conflicts and this can be easily recreated.

**Pull request closed without merging**